### PR TITLE
feat!: support of solid, dotted, dashed styles for polygons, with optimized rendering

### PR DESCRIPTION
**Pull request by monsieurtanuki**
### What
* Now we support dashed style for polygons.
* We use the same algorithm as in polylines, with
  *  `DottedPixelHiker` that returns the visible dots (and only them)
  *  `DashedPixelHiker` that returns the visible segments (and only them)
* The algorithm has been improved in order to take into account side-effects fixed in #1854

### Screenshot
Screenshot_1712493274

### Files
New files:
* `pixel_hiker.dart`: Pixel hikers that list the visible items on the way. Code used to be in `polyline_layer/painter.dart`, but was heavily refactored with #1854 in mind
* `visible_segment.dart`: Cohen-Sutherland algorithm to clip segments as visible into a canvas. Code used to be in `polygon_layer/painter.dart`, and was lightly refactored.

Impacted files:
* `polygon_layer/painter.dart`: now using new file `pixel_hiker.dart` for optimized rendering; moved "clip code" to new file `visible_segment.dart`; minor refactoring about parameter order consistency
* `polyline_layer/painter.dart`: now using new file `pixel_hiker.dart` for optimized rendering; moved "pixel hiker" to new file `pixel_hiker.dart`
* `pages/polygon.dart`: replaced `bool isDotted` with `PolylinePattern pattern` and in one case replaced it with "dashed"
* `polygon_layer/polygon.dart`: BREAKING - replaced `bool isDotted` with `PolylinePattern pattern`
* `polygon_layer/polygon_layer.dart`: minor refactoring
* `polyline_layer/polyline_layer.dart`: minor refactoring

### CODE REVIEW

File: lib/src/layer/polygon_layer/polygon.dart [hunk 1]
```diff
@@ -31,9 +31,11 @@ class Polygon<R extends Object> {
   /// Defaults to false (enabled).
   final bool disableHolesBorder;
 
-  /// Set to true if the border of the [Polygon] should be rendered
-  /// as dotted line.
-  final bool isDotted;
```
Comment (JaffaKetchup) [inline lib/src/layer/polygon_layer/polygon.dart:36]: We need to deprecate this, along with the equivalent in `Polyline`.
I originally did this for `Polyline`s in 72b1a51 after we discussed it, but it was removed in 9c62133 - is there any reason the method I added can't be used?

**Review by JaffaKetchup — COMMENTED** [review]
Untested, going to make some minor file organisation changes, but just just this one comment to consider for now.
This is looking amazing though, and the new painter code is so much simpler!

File: lib/src/layer/polygon_layer/polygon.dart (hunk 1 above)
Comment (monsieurtanuki) [reply lib/src/layer/polygon_layer/polygon.dart:36]: @JaffaKetchup No problem, it's just that in other projects I use the opportunity of (potentially unrelated) "breaking changes" to remove deprecated code almost everywhere: cleaner code.
If you prefer to keep for a moment those deprecated fields, please put them back.

**Comment by monsieurtanuki** [comment]
> Untested, going to make some minor file organisation changes, but just just this one comment to consider for now. This is looking amazing though, and the new painter code is so much simpler!

@JaffaKetchup Thank you! The complexity of the dots or dash patterns multiplied by the "visible?" question asked for those changes.

File: lib/src/layer/general/line_patterns/pixel_hiker.dart [hunk 2]
```diff
@@ -0,0 +1,344 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_map/src/layer/general/line_patterns/stroke_pattern.dart';
+import 'package:meta/meta.dart';
+
+part 'package:flutter_map/src/layer/general/line_patterns/visible_segment.dart';
```
Comment (monsieurtanuki) [inline lib/src/layer/general/line_patterns/pixel_hiker.dart:5]:
```suggestion
part 'visible_segment.dart';
```

**Review by monsieurtanuki — COMMENTED** [review]
@JaffaKetchup I'm fine with all your code organization improvements.

File: lib/src/layer/general/line_patterns/stroke_pattern.dart [hunk 3]
```diff
@@ -15,19 +19,19 @@ class PolylinePattern {
   /// See [spacingFactor] and [PatternFit] for more information about parameters.
   /// [spacingFactor] defaults to 1.5, and [patternFit] defaults to
   /// [PatternFit.scaleUp].
-  const PolylinePattern.dotted({
+  const StrokePattern.dotted({
     double this.spacingFactor = 1.5,
     PatternFit this.patternFit = PatternFit.scaleUp,
   }) : segments = null;
```
Comment (josxha) [inline lib/src/layer/general/line_patterns/stroke_pattern.dart:25]:
```suggestion
  })  : segments = null,
        assert(spacingFactor > 0, 'spacingFactor must be > 0');
```

File: lib/src/layer/general/line_patterns/pixel_hiker.dart [hunk 4]
```diff
@@ -0,0 +1,344 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_map/src/layer/general/line_patterns/stroke_pattern.dart';
```
Comment (josxha) [inline lib/src/layer/general/line_patterns/pixel_hiker.dart:2]:
```suggestion
import 'package:flutter_map/flutter_map.dart';
```

File: lib/src/layer/general/line_patterns/stroke_pattern.dart [hunk 5]
```diff
@@ -1,11 +1,15 @@
-part of 'polyline_layer.dart';
+import 'package:flutter/foundation.dart';
+import 'package:flutter_map/src/layer/polygon_layer/polygon_layer.dart';
+import 'package:flutter_map/src/layer/polyline_layer/polyline_layer.dart';
```
Comment (josxha) [inline lib/src/layer/general/line_patterns/stroke_pattern.dart:3]:
```suggestion
import 'package:flutter_map/flutter_map.dart';
```

**Review by josxha — COMMENTED** [review]
Thanks for the pull request @monsieurtanuki. I did a quick test on web with chrome and didn't experience any problems, nice work! 

The directory `/src/layer/general` makes sense to me, but could we have it's name be in line with `/src/misc` by renaming one of the folders please?

File: lib/src/layer/polygon_layer/polygon.dart (hunk 1 above)
Comment (JaffaKetchup) [reply lib/src/layer/polygon_layer/polygon.dart:36]: Yeah, I suppose that makes more sense. We've discussed internally, and we'll move towards that.

**Comment by monsieurtanuki** [comment]
> The directory `/src/layer/general` makes sense to me, but could we have it's name be in line with `/src/misc` by renaming one of these folders please?

@JaffaKetchup I have no strong opinion on the subject and I'm notoriously neither regarding nor imaginative when it comes to folder and file names, so feel free to move files to a more appropriate folder.

**Review by josxha — APPROVED** [review]
Looks good to me. @JaffaKetchup I assume you want to test the changes yourself before we merge?

**Comment by JaffaKetchup** [comment]
Bit short on time, so if you're happy, that sounds good to me!
Thanks once again @monsieurtanuki :)

**Pull request merged**